orchagent/policerorch.cpp

```diff
@@ -123,119 +142,395 @@ void PolicerOrch::doTask(Consumer &consumer)
+            }
+
+            sai_object_id_t policer_id;
+            sai_status_t status = sai_policer_api->create_policer(
+                    &policer_id, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                SWSS_LOG_ERROR("Failed to create policer %s, rv:%d",
+                        key.c_str(), status);
+                return task_process_status::task_need_retry;
+            }
+
+            SWSS_LOG_NOTICE("Created policer %s", key.c_str());
+            m_syncdPolicers[key].policerOid = policer_id;
+            m_policerRefCounts[key] = 0;
+        }
+        else
+        {
+            auto policer_id = m_syncdPolicers[key].policerOid;
+
+            // The update operation has limitations that it could only update
+            // the rate and the size accordingly.
+            // SR_TCM: CIR, CBS, PBS
+            // TR_TCM: CIR, CBS, PIR, PBS
+            // STORM_CONTROL: CIR, CBS
+            for (auto & attr: attrs)
+            {
+                if (attr.id != SAI_POLICER_ATTR_CBS &&
+                        attr.id != SAI_POLICER_ATTR_CIR &&
+                        attr.id != SAI_POLICER_ATTR_PBS &&
+                        attr.id != SAI_POLICER_ATTR_PIR)
+                {
+                    continue;
+                }
+
+                sai_status_t status = sai_policer_api->set_policer_attribute(
+                        policer_id, &attr);
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_ERROR("Failed to update policer %s attribute, rv:%d",
+                            key.c_str(), status);
+                    return task_process_status::task_need_retry;
+                }
+            }
+            SWSS_LOG_NOTICE("Update policer %s attributes", key.c_str());
+        }
+    }
+    else if (op == DEL_COMMAND)
+    {
+        if (m_syncdPolicers.find(key) == m_syncdPolicers.end())
+        {
+            SWSS_LOG_ERROR("Policer %s does not exists", key.c_str());
+            return task_process_status::task_invalid_entry;
+        }
+
+        if (m_policerRefCounts[key] > 0)
+        {
+            SWSS_LOG_INFO("Policer %s is still referenced", key.c_str());
+            return task_process_status::task_need_retry;
+        }
+
+        sai_status_t status = sai_policer_api->remove_policer(
+                m_syncdPolicers[key].policerOid);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to remove policer %s, rv:%d",
+                    key.c_str(), status);
+            return task_process_status::task_need_retry;
+        }
+
+        SWSS_LOG_NOTICE("Removed policer %s", key.c_str());
+        m_syncdPolicers.erase(key);
+        m_policerRefCounts.erase(key);
+    }
+
+    return task_process_status::task_success;
+}
+
+bool PolicerOrch::isStormControlPolicer(string policer_name)
```

Comment:
This method was introduced as part of supporting another feature to distinguish the policer type based on the name. i.e. if it is an internal policer, the name would start with "_". since BUM storm-control policers are internal, the names start with "_".